llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,242 @@
+}
+
+Value *recursivlyCloneInBB(Value *Op, BasicBlock &BB, ValueToValueMapTy &VMap,
+                           AAResults *AA, DominatorTree *DT, Loop *CurLoop,
+                           MemorySSAUpdater &MSSAU,
+                           bool TargetExecutesOncePerLoop,
+                           SinkAndHoistLICMFlags &Flags) {
+  if (isa<PHINode>(Op)) {
+    LLVM_DEBUG(dbgs() << "Found end: Phi node: do not clone " << *Op << "\n");
+    // return Value that is in the BB, i.e. in the preheader
+    PHINode *PN = dyn_cast<PHINode>(Op);
+    return PN->getIncomingValueForBlock(&BB);
+  }
+
+  Instruction *I = dyn_cast<Instruction>(Op);
+  if (!I) {
+    LLVM_DEBUG(dbgs() << "Found end: " << *Op << "\n");
+    return Op;
+  }
+  if (!canSinkOrHoistInst(*I, AA, DT, CurLoop, MSSAU, TargetExecutesOncePerLoop,
+                          Flags)) {
+    LLVM_DEBUG(dbgs() << "Could not hoist " << *I << "\n");
+    return nullptr;
+  }
+
+  auto NewI = I->clone();
+  NewI->insertBefore(&BB.front());
+  VMap[I] = NewI;
+  LLVM_DEBUG(dbgs() << "Cloning " << *NewI << " into " << BB.getName() << "\n");
+  for (Value *Use : I->operands()) {
+    LLVM_DEBUG(dbgs() << "Try cloning " << *Use << " of Instruction " << *I
+                      << "\n");
+    if (!recursivlyCloneInBB(Use, BB, VMap, AA, DT, CurLoop, MSSAU,
+                             TargetExecutesOncePerLoop, Flags))
+      return nullptr;
+  }
+  llvm::RemapInstruction(NewI, VMap,
+                         RF_NoModuleLevelChanges | RF_IgnoreMissingLocals);
+  return NewI;
+}
+
+// Create the assumption into the Loop Header, that at iteration 0 the condition
+// is true
+void insertMinIterAssumption(ICmpInst &LoopCmpInstr, BasicBlock &LoopPreHeader,
+                             ScalarEvolution &SE, AssumptionCache &AC,
+                             AAResults *AA, DominatorTree *DT,
+                             Loop &CurrentLoop, MemorySSA &MSSA) {
+  MemorySSAUpdater MSSAU(&MSSA);
+  SinkAndHoistLICMFlags LICMFlags(false, CurrentLoop, MSSA);
+
+  Instruction *InsertionPoint = LoopPreHeader.getTerminator();
+
+  // LoopRotate uses SimplifyQuery to determine, if a Branch is conditional or
+  // not. SimplifyQuery can only take an Assumption into account, if it is
+  // before the to-be-evaluated Compare Instruction. Here they are inserted into
+  // the Preheader.
+  IRBuilder<> Builder(dyn_cast<Instruction>(InsertionPoint));
+
+  SCEVExpander Expander(SE, LoopPreHeader.getModule()->getDataLayout(),
+                        "expanded");
+
+  // If AddRecExpr of LHS is available, evaluate at Iteration 0, otherwise
+  // return LHS
+  Value *LHS = expandValueAtZeroIteration(LoopCmpInstr.getOperand(0), SE,
+                                          Expander, InsertionPoint);
+  ValueToValueMapTy VMapLHS;
+  LHS = recursivlyCloneInBB(LHS, LoopPreHeader, VMapLHS, AA, DT, &CurrentLoop,
+                            MSSAU, false, LICMFlags);
+  if (!LHS)
+    return;
```

Comment:
properly delete cloned instructions, if cloning was aborted in the middle of the recursive cloning